juju/juju
# [JUJU-396] juju info/find/download run without a controller

The commands to query CharmHub - info, find download - should be able to be run standalone without needing a running controller. Often, the user just wants to download a charm and inspect it, or find what charms are available etc without having to first bootstrap a controller. One reason for needing a controller was to deal with CharmHub API changes. Given the frequent Juju releases, if there happens to be any backwards incompatible CharmHub changes (which are not expected frequently if at all) , it's easy to release a new Juju client.

All 3 commands now use a consistent way of allowing a different CharmHub (eg beta) backend to be targetted. The commands support a charmhub-url parameter, or you can set the CHARMHUB_URL env var.

One feature not now available with this change is to default to any explicitly configured default series set on the in-scope model. The user is required to use the --series arg if they want to filter on series.

The implementation copies code to deal with parsing the CharmHub API structs from the apiserver package to the CLI package.
The controller client API is deleted and for Juju 3, the apiserver code will be deleted also. There was also other cleanup like deleting the filterFindResults() method which was unused except for a test.

## QA steps

juju info postgresql
juju download mattermost-k8s
juju find mariadb
etc

Check that the output matches what you'd get when running the 2.9 CLI against a controller.
